src/compiler/checker.ts

```diff
@@ -6824,41 +6824,52 @@ namespace ts {
                 getSignaturesOfType(type, SignatureKind.Construct).length === 0;
         }
 
+        function createTransientSymbol(source: Symbol, type: Type) {
+            const symbol = <TransientSymbol>createSymbol(source.flags | SymbolFlags.Transient, source.name);
+            symbol.declarations = source.declarations;
+            symbol.parent = source.parent;
+            symbol.type = type;
+            symbol.target = source;
+            if (source.valueDeclaration) symbol.valueDeclaration = source.valueDeclaration;
+            return symbol;
+        }
+
+        function updateTypeOfMembers(type: Type, update: (propertyType: Type) => Type) {
+            const members: SymbolTable = {};
+            for (const property of getPropertiesOfObjectType(type)) {
+                const original = getTypeOfSymbol(property);
+                const updated = update(original);
+                members[property.name] = updated === original ? property : createTransientSymbol(property, updated);

```

Comment:
I wonder if you could do a copy-on-write sort of thing where you don't need a new map until you create a new symbol - but it's probably not worth the complexity.